gobblin-data-management/src/main/java/gobblin/data/management/trash/Trash.java

```diff
@@ -0,0 +1,269 @@
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.PathFilter;
+import org.apache.hadoop.fs.permission.FsAction;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+import azkaban.utils.Props;
+
+import gobblin.data.management.util.PathUtils;
+
+
+/**
+ * Flexible implementation of Trash similar to Hadoop trash. Allows for injecting cleanup policies for snapshots.
+ */
+public class Trash {
+
+  private static final Logger LOG = LoggerFactory.getLogger(Trash.class);
+  private static final FsPermission PERM =
+      new FsPermission(FsAction.ALL, FsAction.NONE, FsAction.NONE);
+  private static final FsPermission ALL_PERM =
+      new FsPermission(FsAction.ALL, FsAction.ALL, FsAction.ALL);
+
+  public static final String TRASH_LOCATION_KEY = "gobblin.trash.location";
+  public static final String SNAPSHOT_CLEANUP_POLICY_CLASS_KEY = "gobblin.trash.snapshot.cleanup.policy.class";
+  public static final String TRASH_SNAPSHOT_PREFIX = "_TRASH_SNAPSHOT_";
+  public static final String TRASH_IDENTIFIER_FILE = "_THIS_IS_TRASH_DIRECTORY";
+  public static final String DEFAULT_TRASH_DIRECTORY = "_GOBBLIN_TRASH";
+  public static final DateTimeFormatter TRASH_SNAPSHOT_NAME_FORMATTER =
+      DateTimeFormat.forPattern(String.format("'%s'yyyyMMddHHmmss", TRASH_SNAPSHOT_PREFIX)).withZone(DateTimeZone.UTC);
+  public static final PathFilter TRASH_SNAPSHOT_PATH_FILTER = new PathFilter() {
+    @Override
+    public boolean accept(Path path) {
+      return !path.getName().equals(TRASH_IDENTIFIER_FILE) && path.getName().startsWith(TRASH_SNAPSHOT_PREFIX);
+    }
+  };
+  public static final PathFilter TRASH_NOT_SNAPSHOT_PATH_FILTER = new PathFilter() {
+    @Override
+    public boolean accept(Path path) {
+      return !path.getName().equals(TRASH_IDENTIFIER_FILE) && !path.getName().startsWith(TRASH_SNAPSHOT_PREFIX);
+    }
+  };
+
+  /**
+   * Get location of Trash directory. Parsed from props at key {@link #TRASH_LOCATION_KEY}, defaulting to
+   * /home/directory/_GOBBLIN_TRASH.
+   * @param fs {@link org.apache.hadoop.fs.FileSystem} where trash should be found.
+   * @param props {@link azkaban.utils.Props} containing trash configuration.
+   * @return {@link org.apache.hadoop.fs.Path} for trash directory.
+   * @throws java.io.IOException
+   */
+  public static Path getTrashLocation(FileSystem fs, Props props) throws IOException {
+    Preconditions.checkArgument(props.containsKey(TRASH_LOCATION_KEY), "Trash location not provided.");
```

Comment:
Is this check necessary, given that you have a default value for `TRASH_LOCATION_KEY`?